docs/source/data.rst

```diff
@@ -2,11 +2,342 @@ torch.utils.data
+Moreover, it supports
+`both single- and multi-process data loading <Single- and Multi-process Data Loading_>`_,
+as well as automatic `memory pinning <Memory Pinning_>`_.
+
+Dataset Types
+-------------
+
+:class:`~torch.utils.data.DataLoader` supports two different types of datasets:
+
+* `map-style datasets <Map-style datasets_>`_
+
+* `iterable-style datasets <Iterable-style datasets_>`_
+
+Map-style datasets
+^^^^^^^^^^^^^^^^^^
+
+A map-style dataset is one that implements the ``__getitem__`` protocol,
+and represents a map from (possibly non-integral) indices/keys to data samples.
+E.g., such a dataset, when called ``dataset[idx]`` could read and the ``idx``-th
+image and its corresponding label from a folder on the disk.
+
+.. note::
+  :class:`~torch.utils.data.DataLoader` by default constructs a index sampler
+  that yields integral indices.  To make it work with a map-style dataset with
+  non-integral indices/keys, a custom sampler must be provided.
+
+See :class:`~torch.utils.data.Dataset` for more details.
+
+Iterable-style datasets
+^^^^^^^^^^^^^^^^^^^^^^^
+
+An iterable-style dataset is one that implements the ``__iter__`` protocol,
+and represents an iterable over data samples. This type of datasets is
+particularly suitable for cases where random reads are expensive or even
+improbable. E.g., such a dataset, when called ``iter(dataset)``, could return a
+stream of data reading from a database, a remote server, or even logs generated
+in real time.
+
+See :class:`~torch.utils.data.IterableDataset` for more details.
+
+.. note:: When using an :class:`~torch.utils.data.IterableDataset` with
+          `multi-process data loading <Multi-process data loading_>`_. The same
+          dataset object is replicated on each worker process, and thus the
+          replicas must be configured differently to avoid duplicate data. See
+          :class:`~torch.utils.data.IterableDataset` documentations for how to
+          achieve this.
+
+Samplers
+--------
+
+For `map-style datasets <Map-style datasets_>`_, we need a way to specify the
+sequence of indices/keys used in data loading.  The :class:`torch.utils.data.Sampler`
+classes are created for this purpose. They represent iterable objects over the
+indices to map-style datasets.  E.g., in the common case with stochastic
+gradient decent (SGD), a :class:`~torch.utils.data.Sampler` could randomly
+permute a list of indices and yield each one at a time, or yield a small number
+of them for mini-batch SGD.
+
+`Data Loading Strategies`_ section talks about how to use a :class:`~torch.utils.data.Sampler`
+with a :class:`~torch.utils.data.DataLoader`.
+
+Data Loading Strategies
+-----------------------
+
+:class:`~torch.utils.data.DataLoader` constructor receives a
+:attr:`dataset` object as its first argument. Based on the other provided
+arguments, a :class:`~torch.utils.data.DataLoader` operates in one of three
+following strategies:
+
+* `Batched loading from a map-style dataset (default)`_
```

Comment:
It's not the default if you pass in an interable dataset, so I don't think we should mention that, unless you divide those points into two categories.